pkg/otelcollector/infra-meter/infra-meter.go

```diff
@@ -70,19 +73,88 @@ func addInfraMeter(
 			return fmt.Errorf("invalid id %q: %w", origName, err)
 		}
-		id = component.NewIDWithName(id.Type(), normalizeComponentName(pipelineName, id.Name()))
+
+		configSHA := sha256.New()
+		if _, err := configSHA.Write([]byte(receiverConfig.String())); err != nil {
+			return fmt.Errorf("failed to prepare sha256 for receiver '%s' in infra_meter: '%s' of Policy: '%s': %w", origName, infraMeterName, policyName, err)
+		}
+		id = component.NewIDWithName(id.Type(), fmt.Sprintf("%x", configSHA.Sum(nil)))
+		strID := id.String()
+
+		// If receiver is already present with given id and per-agent-group = false, skip adding receiver with per-agent-group = true.
+		if _, ok := config.Receivers[strID]; ok && infraMeter.PerAgentGroup {
+			receiverIDs[origName] = strID
+			continue
+		}
+
 		var cfg any
 		cfg = receiverConfig.AsMap()
 		id, cfg = leaderonlyreceiver.WrapConfigIf(infraMeter.PerAgentGroup, id, cfg)
-		receiverIDs[origName] = id.String()
-		config.AddReceiver(id.String(), cfg)
+		strID = id.String()
+
+		// Remove receiver for per-agent-group infra-meter if a receiver with the same config already exists without per-agent-group.
+		if !infraMeter.PerAgentGroup {
+			updatedID, _ := leaderonlyreceiver.WrapConfigIf(true, id, cfg)
+			if _, ok := config.Receivers[updatedID.String()]; ok {
+				delete(config.Receivers, updatedID.String())
+				for key, value := range receiverIDs {
+					if value == updatedID.String() {
+						receiverIDs[key] = id.String()
+					}
+				}
+
+				// Update pipelines to use the new receiver id.
+				for key, value := range config.Service.Pipelines {
+					if slices.Contains(value.Receivers, updatedID.String()) {
+						value.Receivers[slices.Index(value.Receivers, updatedID.String())] = id.String()
+						config.Service.Pipelines[key] = value
+					}
+				}
+			}
+		}
+		receiverIDs[origName] = strID
+		config.AddReceiver(strID, cfg)
 	}
 
 	for origName, processorConfig := range infraMeter.Processors {
-		id := normalizeComponentName(pipelineName, origName)
-		processorIDs[origName] = id
-		var cfg any = processorConfig.AsMap()
-		config.AddProcessor(id, cfg)
+		var id component.ID
+		if err := id.UnmarshalText([]byte(origName)); err != nil {
+			return fmt.Errorf("invalid id %q: %w", origName, err)
+		}
+
+		selectorsList := []interface{}{}
+		var selectors *structpb.Value
+		if id.Type() == otelconsts.ProcessorK8sAttributes {
+			var ok bool
+			selectors, ok = processorConfig.Fields[otelconsts.ProcessorK8sAttributesSelectors]
+			if ok && selectors != nil {
+				selectorsList = selectors.GetListValue().AsSlice()
+				delete(processorConfig.Fields, otelconsts.ProcessorK8sAttributesSelectors)
+			}
+		}
+
+		configSHA := sha256.New()
+		if _, err := configSHA.Write([]byte(processorConfig.String())); err != nil {
+			return fmt.Errorf("failed to prepare sha256 for receiver '%s' in infra_meter: '%s' of Policy: '%s': %w", origName, infraMeterName, policyName, err)
```

Comment:
sha.Write never fails. You can safely panic / ignore error

Reply:
golangci-lint was complaining about it so I kept it.

Reply:
btw, sha256 defines helper function https://pkg.go.dev/crypto/sha256#Sum256, this should simplify

Reply:
Made the changes using their helper